test/cpp/qps/client_async.cc

```diff
@@ -949,10 +949,182 @@ std::unique_ptr<Client> CreateAsyncClient(const ClientConfig& config) {
       return nullptr;
   }
 }
+
 std::unique_ptr<Client> CreateGenericAsyncStreamingClient(
     const ClientConfig& args) {
   return std::unique_ptr<Client>(new GenericAsyncStreamingClient(args));
 }
 
+struct CallbackClientRpcData {
+  CallbackClientRpcData()
+      : thread_mu_(), thread_cv_(), rpc_done_(false), alarm_(nullptr) {}
+
+  ~CallbackClientRpcData() { delete alarm_; }
+
+  std::mutex thread_mu_;
+  std::condition_variable thread_cv_;
+  bool rpc_done_;
+  Alarm* alarm_;
+};
+
+class CallbackClient
+    : public ClientImpl<BenchmarkService::Stub, SimpleRequest> {
+ public:
+  CallbackClient(const ClientConfig& config)
+      : ClientImpl<BenchmarkService::Stub, SimpleRequest>(
+            config, BenchmarkStubCreator) {
+    num_threads_ = NumThreads(config);
+    ctxs_.resize(num_threads_);
+    threads_done_ = 0;
+    SetupLoadTest(config, num_threads_);
+  }
+
+  virtual ~CallbackClient() {}
+
+  virtual bool InitThreadFuncImpl(size_t thread_idx) { return true; }
+
+  virtual bool ThreadFuncImpl(Thread* t, size_t thread_idx) = 0;
+
+ protected:
+  size_t num_threads_;
+  std::mutex mu_;
+  std::condition_variable cv_;
+  std::vector<CallbackClientRpcData*> ctxs_;
+  size_t threads_done_;
+
+  void NotifyThread(size_t thread_idx) {
+    std::mutex* thread_mu = &ctxs_[thread_idx]->thread_mu_;
+    std::condition_variable* thread_cv = &ctxs_[thread_idx]->thread_cv_;
+    std::lock_guard<std::mutex> thread_lock(*thread_mu);
+    ctxs_[thread_idx]->rpc_done_ = true;
+    thread_cv->notify_one();
+  }
+
+  void IssueCallbackRpc(size_t thread_idx, Thread* t) {
+    if (!InitThreadFuncImpl(thread_idx)) {
```

Comment:
So InitThreadImpl is just a function from client.h that we override. It really does not involve in the synchronization. The only place the rpc_done_ is set true is under the final callback when ThreadCompleted is true. This variable as you mentioned earlier is incorrectly named and should be more on the lines of thread_done_. My upcoming commit however should hopefully remove the need of using this.

Reply:
It seems to me like it is involved in synchronization by nature of being called in this method. `IssueCallbackRpc` is called by `ThreadFunc`. Let's say at some point we decide to implement InitThreadFuncImpl to return false sometimes, and it happens to return false when `ThreadFunc` calls it. Then `ThreadFunc` will call `WaitForCompletionOfAllRpcs`, which will wait on the condition: `!ctxs_[thread_idx]->rpc_done_`. But that condition is never true, because "the only place the rpc_done_ is set true is under the final callback when ThreadCompleted is true": ThreadFuncImpl is never called, so the final callback, or indeed any callback, never happens. Does that make sense?